### PR TITLE
Add InputSet in build_config

### PR DESCRIPTION
**Pull request by natebosch**
This gives a more standard way to express include/exclude pairs and a
more flexible way to parse them.

### CODE REVIEW

File: build_config/lib/src/build_config.dart [hunk 1]
```diff
@@ -12,6 +12,37 @@ import 'package:path/path.dart' as p;
 import 'parse.dart';
 import 'pubspec.dart';
 
+/// A glob filter on files inputs or sources.
```
Comment (jakemac53) [inline build_config/lib/src/build_config.dart]: nit: wording is a bit off

File: build_config/lib/src/build_config.dart [hunk 2]
```diff
@@ -12,6 +12,37 @@ import 'package:path/path.dart' as p;
...
+  /// The globs to include in the set.
+  ///
+  /// May be null or empty which means every possible path (like `'**'`).
+  final List<String> include;
+
+  /// The globs as a subset of [include] to remove from the set.
+  ///
+  /// May be null or empty which means every path in [include].
+  final List<String> exclude;
+  const InputSet({this.include, this.exclude});
```
Comment (jakemac53) [inline build_config/lib/src/build_config.dart]: nit: newline above

File: build_config/lib/src/build_config.dart [hunk 3]
```diff
@@ -12,6 +12,37 @@ import 'package:path/path.dart' as p;
...
+
+  @override
+  String toString() {
+    final result = new StringBuffer()..write('paths matching $include');
```
Comment (jakemac53) [inline build_config/lib/src/build_config.dart]: I don't think you want the `write` call here? (looks like its handled better right below)

File: build_config/lib/src/parse.dart [hunk 4]
```diff
@@ -322,3 +315,33 @@ List<String> _readListOfStringsOrThrow(
...
+  }
+
+  if (value is Map) {
+    final invalidOptions = value.keys.toList()
+      ..removeWhere((k) => k == 'include' || k == 'exclude');
+    if (invalidOptions.isNotEmpty) {
+      throw new ArgumentError('Got invalid options `$invalidOptions` '
+          'for $option. Only "include", and "exclude" are allowed');
+    }
+    final include = value.containsKey('include')
```
Comment (jakemac53) [inline build_config/lib/src/parse.dart]: You could simplify these a bit:

```dart
final include = (value['include'] as List<String>)?.toList()
```

File: build_config/lib/src/build_config.dart (hunk 1 above)
Comment (natebosch) [reply build_config/lib/src/build_config.dart]: Done

File: build_config/lib/src/build_config.dart (hunk 2 above)
Comment (natebosch) [reply build_config/lib/src/build_config.dart]: Done

File: build_config/lib/src/build_config.dart (hunk 3 above)
Comment (natebosch) [reply build_config/lib/src/build_config.dart]: yup, broken refactor. Fixed.

File: build_config/lib/src/parse.dart (hunk 4 above)
Comment (natebosch) [reply build_config/lib/src/parse.dart]: The suggested `as` won't work in Dart 2 semantics so I went with the future-proof approach.

**Pull request merged**